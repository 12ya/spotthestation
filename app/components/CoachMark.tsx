import React from "react"
import { ViewStyle, View, PressableProps, TextStyle, StyleProp } from "react-native"
import { Button, Icon, IconTypes, Text } from "."
import { TxKeyPath } from "../i18n"
import { typography } from "../theme"
import { colors } from "../theme/colors"
import { StyleFn, useStyles } from "../utils/useStyles"

export interface CoachMarkProps {
  icon?: IconTypes
  title?: TxKeyPath
  bodyText?: TxKeyPath
  stage?: number
  style?: StyleProp<ViewStyle>
  onPressNext?: PressableProps["onPress"]
  onPressFinish?: PressableProps["onPress"]
  totalStages?: number
  arrowStyle?: StyleProp<ViewStyle>
}

export function CoachMark({
  icon,
  title,
  bodyText,
  stage,
  style,
  onPressNext,
  onPressFinish,
  totalStages,
  arrowStyle,
}: CoachMarkProps) {
  const {
    $modalBodyContainer,
    $buttonsContainer,
    $skipButton,
    $skipButtonText,
    $nextButton,
    $nextButtonText,
    $title,
    $body,
    $stage,
  } = useStyles(styles)

  const renderArrow = () => {
    return (
      <View style={arrowStyle}>
        <Icon icon="bigArrow" size={55} />
      </View>
    )
  }

  return (
    <View
      accessible
      accessibilityLabel="coach mark"
      accessibilityHint="coach mark"
      accessibilityRole="text"
      style={[$modalBodyContainer, style]}
    >
      {renderArrow()}
      <Text text={`${stage}/${totalStages}`} style={$stage} />
      {Boolean(icon) && <Icon icon={icon} size={44} />}
      <Text tx={title} style={$title} />
      <Text tx={bodyText} style={$body} />
      {stage === totalStages ? (
        <Button
          accessible
          accessibilityLabel="finish button"
          accessibilityHint="finish coach mark"
          tx="homeScreen.coachMarks.finish"
          textStyle={$nextButtonText}
          style={$nextButton}
          pressedStyle={$nextButton}
          onPress={onPressFinish}
        />
      ) : (
        <View style={$buttonsContainer}>
          <Button
            accessible
            accessibilityLabel="skip button"
            accessibilityHint="skip coach mark"
            tx="homeScreen.coachMarks.skip"
            textStyle={$skipButtonText}
            style={$skipButton}
            pressedStyle={$skipButton}
            onPress={onPressFinish}
          />
          <Button
            accessible
            accessibilityLabel="next button"
            accessibilityHint="next coach mark"
            tx="homeScreen.coachMarks.next"
            textStyle={$nextButtonText}
            style={$nextButton}
            pressedStyle={$nextButton}
            onPress={onPressNext}
          />
        </View>
      )}
    </View>
  )
}

const styles: StyleFn = ({ scale, fontSizes, lineHeights }) => {
  const $modalBodyContainer: ViewStyle = {
    backgroundColor: colors.palette.buttonBlue,
    borderRadius: scale(16),
    alignItems: "center",
    paddingVertical: 36,
    paddingHorizontal: 30,
    width: "100%",
  }

  const $buttonsContainer: ViewStyle = {
    flexDirection: "row",
    justifyContent: "space-between",
    width: "100%",
    alignSelf: "flex-end",
  }

  const $skipButton: ViewStyle = {
    backgroundColor: "transparent",
    borderWidth: 0,
    height: scale(56),
    minWidth: scale(140),
  }

  const $skipButtonText: TextStyle = {
    fontFamily: typography.primary.medium,
    fontSize: fontSizes[18],
    lineHeight: lineHeights[22],
    color: colors.palette.neutral100,
  }

  const $nextButton: ViewStyle = {
    height: scale(56),
    backgroundColor: colors.palette.neutral100,
    borderRadius: scale(28),
    borderWidth: 0,
    minWidth: scale(140),
  }

  const $nextButtonText: TextStyle = {
    fontFamily: typography.primary.medium,
    fontSize: fontSizes[18],
    lineHeight: lineHeights[22],
    color: colors.palette.buttonBlue,
  }

  const $title: TextStyle = {
    fontFamily: typography.primary.normal,
    fontSize: fontSizes[24],
    lineHeight: lineHeights[29],
    color: colors.palette.neutral100,
    paddingBottom: 12,
    paddingTop: 18,
  }

  const $body: TextStyle = {
    fontFamily: typography.primary.normal,
    fontSize: fontSizes[18],
    lineHeight: lineHeights[22],
    color: colors.palette.neutral100,
    paddingBottom: 36,
  }

  const $stage: TextStyle = {
    fontFamily: typography.primary.normal,
    fontSize: fontSizes[16],
    lineHeight: lineHeights[20],
    color: colors.palette.neutral100,
    position: "absolute",
    top: 0,
    left: 0,
    padding: 24,
    opacity: 0.5,
  }

  return {
    $modalBodyContainer,
    $buttonsContainer,
    $skipButton,
    $skipButtonText,
    $nextButton,
    $nextButtonText,
    $title,
    $body,
    $stage,
  }
}
