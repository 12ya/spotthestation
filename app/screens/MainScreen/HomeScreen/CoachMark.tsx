import React from "react"
import { ViewStyle, View, PressableProps, TextStyle } from "react-native"
import { Button, Icon, IconTypes, Text } from "../../../components"
import { TxKeyPath } from "../../../i18n"
import { fontSizes, lineHeights, scale, typography } from "../../../theme"
import { colors } from "../../../theme/colors"
import { normalizeHeight } from "../../../utils/normalizeHeight"

export interface CoachMarkProps {
  icon?: IconTypes,
  title?: TxKeyPath,
  bodyText?: TxKeyPath,
  stage?: number,
  style?: ViewStyle,
  onPressNext?: PressableProps["onPress"]
  onPressFinish?: PressableProps["onPress"]
}

export function CoachMark({ icon, title, bodyText, stage, style, onPressNext, onPressFinish }: CoachMarkProps) {
  const renderArrow = () => {
    switch (stage) {
      case 1: return <View style={$location}><Icon icon="bigArrow" size={55} /></View>
      case 2: return <View style={$sightings}><Icon icon="bigArrow" size={55} /></View>
      case 3: return <View style={$globe}><Icon icon="bigArrow" size={55} /></View>
      case 4: return <View style={$map}><Icon icon="bigArrow" size={55} /></View>
      case 5: return <View style={$navigation}><Icon icon="bigArrow" size={55} /></View>
      default: return ''
    }
  }

  return (
    <View 
      accessible
      accessibilityLabel="coach mark"
      accessibilityHint="coach mark"
      accessibilityRole="text"
      style={[$modalBodyContainer, style]}
    >
      {renderArrow()}
      <Text text={`${stage}/5`} style={$stage} />
      <Icon icon={icon} size={44} />
      <Text tx={title} style={$title} />
      <Text tx={bodyText} style={$body} />
      {stage === 5 ? <Button
          accessible
          accessibilityLabel="finish button"
          accessibilityHint="finish coach mark"
          tx="homeScreen.coachMarks.finish"
          textStyle={$nextButtonText}
          style={$nextButton}
          pressedStyle={$nextButton}
          onPress={onPressFinish}
        /> : <View style={$buttonsContainer}>
        <Button
          accessible
          accessibilityLabel="skip button"
          accessibilityHint="skip coach mark"
          tx="homeScreen.coachMarks.skip"
          textStyle={$skipButtonText}
          style={$skipButton}
          pressedStyle={$skipButton}
          onPress={onPressFinish}
        />
        <Button
          accessible
          accessibilityLabel="next button"
          accessibilityHint="next coach mark"
          tx="homeScreen.coachMarks.next"
          textStyle={$nextButtonText}
          style={$nextButton}
          pressedStyle={$nextButton}
          onPress={onPressNext}
        />
      </View>}
    </View>
  )
}

const $modalBodyContainer: ViewStyle = {
  backgroundColor: colors.palette.buttonBlue,
  borderRadius: scale(16),
  alignItems: 'center',
  paddingVertical: 36,
  paddingHorizontal: 30,
  width: '100%',
}

const $buttonsContainer: ViewStyle = {
  flexDirection: 'row',
  justifyContent: 'space-between',
  width: '100%',
  alignSelf: 'flex-end'
}

const $skipButton: ViewStyle = {
  backgroundColor: 'transparent',
  borderWidth: 0,
  height: scale(56),
  minWidth: scale(140)
}

const $skipButtonText: TextStyle = {
  fontFamily: typography.primary.medium,
  fontSize: fontSizes[18],
  lineHeight: lineHeights[22],
  color: colors.palette.neutral100,
}

const $nextButton: ViewStyle = {
  height: scale(56),
  backgroundColor: colors.palette.neutral100,
  borderRadius: scale(28),
  borderWidth: 0,
  minWidth: scale(140)
}

const $location: ViewStyle = {
  position: "absolute",
  top: -scale(80),
  right: -scale(5),
}
const $sightings: ViewStyle = {
  position: "absolute",
  top: -scale(80),
}
const $globe: ViewStyle = {
  position: "absolute",
  top: -normalizeHeight(.2),
  right: 0,
  transform: [{ rotate: "-130deg" }]
}
const $map: ViewStyle = {
  position: "absolute",
  bottom: -scale(80),
  transform: [{ rotate: "180deg" }]
}
const $navigation: ViewStyle = {
  position: "absolute",
  bottom: -scale(80),
  transform: [{ rotate: "180deg" }]
}

const $nextButtonText: TextStyle = {
  fontFamily: typography.primary.medium,
  fontSize: fontSizes[18],
  lineHeight: lineHeights[22],
  color: colors.palette.buttonBlue,
}

const $title: TextStyle = {
  fontFamily: typography.primary.normal,
  fontSize: fontSizes[24],
  lineHeight: lineHeights[29],
  color: colors.palette.neutral100,
  paddingBottom: 12,
  paddingTop: 18
}

const $body: TextStyle = {
  fontFamily: typography.primary.normal,
  fontSize: fontSizes[18],
  lineHeight: lineHeights[22],
  color: colors.palette.neutral100,
  paddingBottom: 36
}

const $stage: TextStyle = {
  fontFamily: typography.primary.normal,
  fontSize: fontSizes[16],
  lineHeight: lineHeights[20],
  color: colors.palette.neutral100,
  position: "absolute",
  top: 0,
  left: 0,
  padding: 24,
  opacity: 0.5
}
