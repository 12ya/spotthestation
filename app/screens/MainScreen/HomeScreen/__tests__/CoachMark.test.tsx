/* eslint-disable react/display-name */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import React from "react"
import { CoachMark } from "../CoachMark"
import { render } from "@testing-library/react-native"

it("renders correctly", () => {
  const tree = render(
    <CoachMark
      icon="mapPinOutlined"
      title="homeScreen.coachMarks.locationTitle"
      bodyText="homeScreen.coachMarks.locationData"
      style={$style}
      stage={0}
      onPressFinish={jest.fn()}
      onPressNext={jest.fn()}
    />,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

const $style = {
  marginTop: 18,
}
