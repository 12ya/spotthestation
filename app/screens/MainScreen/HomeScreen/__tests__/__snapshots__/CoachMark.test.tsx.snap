// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessibilityHint="coach mark"
  accessibilityLabel="coach mark"
  accessibilityRole="text"
  accessible={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#0B3D91",
        "borderRadius": 23,
        "paddingHorizontal": 30,
        "paddingVertical": 36,
        "width": "100%",
      },
      Object {
        "marginTop": 18,
      },
    ]
  }
>
  <div
    style={
      Object {
        "color": "#FFFFFF",
        "fontFamily": "Helvetica Neue",
        "fontSize": 23,
        "left": 0,
        "lineHeight": 29,
        "opacity": 0.5,
        "padding": 24,
        "position": "absolute",
        "top": 0,
      }
    }
  >
    0/5
  </div>
  <div />
  <div
    style={
      Object {
        "color": "#FFFFFF",
        "fontFamily": "Helvetica Neue",
        "fontSize": 35,
        "lineHeight": 42,
        "paddingBottom": 12,
        "paddingTop": 18,
      }
    }
  />
  <div
    style={
      Object {
        "color": "#FFFFFF",
        "fontFamily": "Helvetica Neue",
        "fontSize": 26,
        "lineHeight": 32,
        "paddingBottom": 36,
      }
    }
  />
  <View
    style={
      Object {
        "alignSelf": "flex-end",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
  >
    <div />
    <div />
  </View>
</View>
`;
